File: samples/sample.cs
/**
 * C# sample demonstrating all token types and syntax.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.IO;
using static System.Math;

// Alias
using StringList = System.Collections.Generic.List<string>;

// Comments
// Single line comment
/* Multi-line
   comment */
/// <summary>
/// XML documentation comment
/// </summary>

namespace MyNamespace
{
    // Numeric literals
    class LiteralExamples
    {
        int decimal_int = 1234;
        int hex_int = 0xDEADBEEF;
        int binary_int = 0b1010_1010;
        long long_num = 1000L;
        ulong ulong_num = 1000UL;
        float float_num = 3.14f;
        double double_num = 2.718;
        decimal decimal_num = 9.99m;
        double scientific = 1.5e-3;
        
        // String literals
        string regular = "string";
        string verbatim = @"C:\path\to\file
multiline";
        string interpolated = $"Value: {decimal_int}";
        string interpolatedVerbatim = $@"Raw: {decimal_int}";
        
        // Character literals
        char charLit = 'A';
        char escaped = '\n';
        char unicode = '\u0041';
        
        // Boolean
        bool isTrue = true;
        bool isFalse = false;
        
        // Null
        object nullValue = null;
    }
    
    // Enums
    public enum Color { Red, Green, Blue }
    public enum Status : byte { Idle = 0, Running = 1, Stopped = 2 }
    
    [Flags]
    public enum Permissions : int { Read = 1, Write = 2, Execute = 4 }
    
    // Structures
    public struct Point
    {
        public int X { get; set; }
        public int Y { get; set; }
        
        public Point(int x, int y)
        {
            X = x;
            Y = y;
        }
    }
    
    // Interfaces
    public interface IAnimal
    {
        void Speak();
        int Age { get; set; }
    }
    
    public interface IMovable
    {
        void Move(int distance);
    }
    
    // Abstract class
    public abstract class Animal : IAnimal
    {
        public string Name { get; set; }
        public int Age { get; set; }
        
        protected Animal(string name, int age)
        {
            Name = name;
            Age = age;
        }
        
        public virtual void Speak()
        {
            Console.WriteLine($"{Name} makes a sound");
        }
        
        public abstract void MakeNoise();
    }
    
    // Regular class
    public class Dog : Animal, IMovable
    {
        private string breed;
        
        // Auto-properties
        public string Breed
        {
            get { return breed; }
            set { breed = value; }
        }
        
        // Expression-bodied property
        public string BreedInfo => $"Breed: {breed}";
        
        // Backing field
        private int _age;
        public new int Age
        {
            get { return _age; }
            set { _age = value; }
        }
        
        // Init-only property (C# 9.0)
        public string Owner { get; init; }
        
        // Required property (C# 11.0)
        public required string Id { get; set; }
        
        // Constructor
        public Dog(string name, string breed) : base(name, 0)
        {
            this.breed = breed;
        }
        
        // Constructor chaining
        public Dog(string name, string breed, int age) : this(name, breed)
        {
            Age = age;
        }
        
        // Method
        public override void Speak()
        {
            Console.WriteLine("Woof!");
        }
        
        public override void MakeNoise()
        {
            Console.WriteLine("Bark bark");
        }
        
        // Expression-bodied method
        public void Move(int distance) => Console.WriteLine($"Moved {distance} meters");
        
        // Sealed override
        public sealed override void Speak() => base.Speak();
        
        // Static method
        public static Dog CreateDefault()
        {
            return new Dog("Buddy", "Labrador");
        }
        
        // Static constructor
        static Dog()
        {
            // Static initialization
        }
        
        // Operator overloading
        public static bool operator ==(Dog left, Dog right)
        {
            return left.Name == right.Name;
        }
        
        public static bool operator !=(Dog left, Dog right)
        {
            return !(left == right);
        }
        
        public static Dog operator +(Dog d1, Dog d2)
        {
            return new Dog($"{d1.Name}+{d2.Name}", "Mixed");
        }
        
        // Conversion operators
        public static explicit operator string(Dog dog)
        {
            return dog.Name;
        }
        
        public static implicit operator Dog(string name)
        {
            return new Dog(name, "Unknown");
        }
        
        // Indexer
        public string this[int index]
        {
            get { return $"Item{index}"; }
            set { }
        }
    }
    
    // Generic class
    public class Container<T> where T : class
    {
        private List<T> items = new List<T>();
        
        public void Add(T item) => items.Add(item);
        public T Get(int index) => items[index];
        public IEnumerable<T> GetAll() => items;
    }
    
    // Generic with constraints
    public class Repository<T> where T : IAnimal, new()
    {
        public T Create() => new T();
    }
    
    // Record class (C# 9.0)
    public record PersonRecord(string Name, int Age);
    
    public record Dog_Record(string Name, string Breed) : Animal(Name, 0)
    {
        public override void MakeNoise() => Console.WriteLine("Woof");
    }
    
    // Anonymous type (created inline)
    var anonymous = new { Name = "John", Age = 30 };
    
    // Delegate
    public delegate void AnimalEventHandler(Animal animal);
    
    // Event
    public class AnimalShelter
    {
        public event AnimalEventHandler AnimalArrived;
        
        public void Register(Animal animal)
        {
            AnimalArrived?.Invoke(animal);
        }
    }
    
    // Lambda and expression trees
    public class ExpressionExamples
    {
        Func<int, int> square = x => x * x;
        Func<int, int, int> add = (x, y) => x + y;
        Func<int, int, int> multiline = (x, y) =>
        {
            int result = x + y;
            return result * 2;
        };
        
        Action<string> print = msg => Console.WriteLine(msg);
        
        Predicate<int> isEven = n => n % 2 == 0;
    }
    
    // Extension methods
    public static class StringExtensions
    {
        public static string Reverse(this string str)
        {
            return new string(str.Reverse().ToArray());
        }
    }
    
    // Partial class
    public partial class PartialExample
    {
        public void Method1() { }
    }
    
    public partial class PartialExample
    {
        public void Method2() { }
    }
    
    // Control flow
    public class ControlFlow
    {
        public void Example(int x)
        {
            // if-else
            if (x > 0)
            {
                Console.WriteLine("Positive");
            }
            else if (x < 0)
            {
                Console.WriteLine("Negative");
            }
            else
            {
                Console.WriteLine("Zero");
            }
            
            // Switch expression (C# 8.0)
            string result = x switch
            {
                > 0 => "Positive",
                < 0 => "Negative",
                _ => "Zero"
            };
            
            // Traditional switch
            switch (x)
            {
                case 1:
                    Console.WriteLine("One");
                    break;
                case 2:
                case 3:
                    Console.WriteLine("Two or three");
                    break;
                default:
                    break;
            }
            
            // for loop
            for (int i = 0; i < 10; i++)
            {
                if (i > 5) break;
                if (i == 3) continue;
                Console.WriteLine(i);
            }
            
            // foreach loop
            var numbers = new[] { 1, 2, 3, 4, 5 };
            foreach (int num in numbers)
            {
                Console.WriteLine(num);
            }
            
            // while loop
            while (true)
            {
                break;
            }
            
            // do-while loop
            do
            {
                Console.WriteLine("At least once");
            } while (false);
        }
    }
    
    // Exception handling
    public class ExceptionHandling
    {
        public void Example()
        {
            try
            {
                throw new InvalidOperationException("Error");
            }
            catch (ArgumentNullException ex)
            {
                Console.WriteLine($"Null argument: {ex.Message}");
            }
            catch (InvalidOperationException ex) when (ex.Message.Contains("Error"))
            {
                Console.WriteLine("Caught specific error");
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                Console.WriteLine("Cleanup");
            }
        }
    }
    
    // LINQ examples
    public class LINQExamples
    {
        public void Example()
        {
            var numbers = new[] { 1, 2, 3, 4, 5, 6 };
            
            // Method syntax
            var result1 = numbers.Where(x => x > 2)
                                 .Select(x => x * 2)
                                 .ToList();
            
            // Query syntax
            var result2 = from num in numbers
                         where num > 2
                         select num * 2;
            
            // Complex query
            var result3 = from num in numbers
                         where num > 2
                         group num by num % 2 into g
                         select new { Remainder = g.Key, Count = g.Count() };
            
            // Let clause
            var result4 = from num in numbers
                         let squared = num * num
                         where squared > 10
                         select squared;
            
            // Join
            var result5 = from num in numbers
                         join n in numbers on num equals n into g
                         select new { num, Count = g.Count() };
        }
    }
    
    // Async/await
    public class AsyncExamples
    {
        public async Task SimpleAsync()
        {
            await Task.Delay(1000);
            Console.WriteLine("Done");
        }
        
        public async Task<int> ReturnValue()
        {
            await Task.Delay(500);
            return 42;
        }
        
        public async Task MultipleAwaits()
        {
            var task1 = Task.Delay(100);
            var task2 = Task.Delay(200);
            
            await task1;
            await task2;
            
            await Task.WhenAll(task1, task2);
            await Task.WhenAny(task1, task2);
        }
    }
    
    // Attributes
    [Serializable]
    [Obsolete("Use NewClass instead", false)]
    public class AttributeExample
    {
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Style", "")]
        public void Method()
        {
        }
    }
    
    // Nullable reference types (C# 8.0)
    public class NullableExample
    {
        public string? NullableString { get; set; }
        public string NonNullableString { get; set; } = "";
        
        public void Method(string? value)
        {
            if (value is not null)
            {
                Console.WriteLine(value);
            }
        }
    }
    
    // Pattern matching (C# 7.0+)
    public class PatternMatching
    {
        public void Example(object obj)
        {
            // Type pattern
            if (obj is Dog dog)
            {
                Console.WriteLine(dog.Breed);
            }
            
            // Property pattern
            if (obj is Dog { Age: > 5 })
            {
                Console.WriteLine("Old dog");
            }
            
            // Relational pattern
            if (obj is int { } x && x > 10)
            {
                Console.WriteLine("Large number");
            }
            
            // List pattern (C# 11.0)
            if (obj is [1, 2, ..])
            {
                Console.WriteLine("Starts with 1, 2");
            }
        }
    }
    
    // Operators - Arithmetic
    public void ArithmeticOps(int a, int b)
    {
        int result = a + b;
        result = a - b;
        result = a * b;
        result = a / b;
        result = a % b;
        a++;
        b--;
        a += 5;
        b -= 5;
    }
    
    // Operators - Bitwise
    public void BitwiseOps(int a, int b)
    {
        int result = a & b;
        result = a | b;
        result = a ^ b;
        result = ~a;
        result = a << 1;
        result = a >> 1;
        a &= b;
        a |= b;
    }
    
    // Operators - Logical
    public void LogicalOps(bool a, bool b)
    {
        bool result = a && b;
        result = a || b;
        result = !a;
        result = a == b;
        result = a != b;
        result = a < b;
        result = a > b;
    }
    
    // Null-coalescing and null-conditional
    public void NullOps(string? str, Dog? dog)
    {
        string value = str ?? "default";
        int? length = str?.Length;
        string? breed = dog?.Breed;
        string result = dog?.Breed ?? "Unknown";
    }
    
    // Using statements (C# 8.0+)
    public void UsingDeclaration()
    {
        using var file = new FileStream("test.txt", FileMode.Open);
        // File is automatically disposed
    }
    
    // Using declarations (traditional)
    public void TraditionalUsing()
    {
        using (var file = new FileStream("test.txt", FileMode.Open))
        {
            // Use file
        }
    }
    
    // Tuple
    public (string Name, int Age) GetPerson()
    {
        return ("John", 30);
    }
    
    public void DeconstructTuple()
    {
        var (name, age) = GetPerson();
    }
    
    // Comments with special syntax
    // TODO: Complete this
    // FIXME: Fix the bug
    // NOTE: Important
    // HACK: Temporary fix
}

class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Hello, World!");
    }
}

File: samples/sample.cpp
/**
 * C++ sample demonstrating all token types and syntax.
 */

#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <memory>
#include <algorithm>
#include <functional>

// Single line comment
/* Multi-line
   comment */

using namespace std;

// Preprocessor directives
#define MACRO_NAME 42
#define FUNCTION_MACRO(x) ((x) * 2)
#ifdef DEBUG
#define DEBUG_PRINT(msg) cout << msg << endl
#endif
#pragma once

// Namespace
namespace MyNamespace {
    // Constants and literals
    const int CONSTANT_INT = 100;
    constexpr int COMPILE_TIME_INT = 200;
    const double PI = 3.14159;
    const char* STRING_LITERAL = "string";
    const wchar_t WIDE_CHAR = L'A';
    const char UTF8[] = u8"UTF-8";
    const char16_t UTF16[] = u"UTF-16";
    const char32_t UTF32[] = U"UTF-32";
    
    // Numeric literals
    int decimal = 1234;
    int hex = 0xDEADBEEF;
    int octal = 0755;
    int binary = 0b1010;
    long longNum = 1000L;
    unsigned int unsignedInt = 100U;
    float floatNum = 3.14f;
    double doubleNum = 2.718;
    double scientific = 1.5e-3;
    
    // String literals
    string str1 = "C-style string";
    string str2 = "escaped\nnewline\ttab";
    string raw = R"(raw string\n no escape)";
    
    // Boolean
    bool flag = true;
    bool notFlag = false;
    
    // Null pointer
    int* nullPtr = nullptr;
    
    // Enumerations
    enum Color { RED, GREEN, BLUE };
    enum class Status : unsigned char { IDLE = 0, RUNNING = 1, STOPPED = 2 };
    
    // Type aliases
    using IntVector = vector<int>;
    using StringMap = map<string, string>;
    typedef unsigned long ulong;
    
    // Structures
    struct Point {
        int x;
        int y;
        
        Point() : x(0), y(0) {}
        Point(int x, int y) : x(x), y(y) {}
    };
    
    // Union
    union Data {
        int i;
        float f;
        char c;
    };
    
    // Classes
    class Animal {
    private:
        string name;
        mutable int callCount = 0;
        
    protected:
        int age;
        
    public:
        // Constructors
        Animal() = default;
        explicit Animal(const string& name) : name(name) {}
        Animal(const Animal&) = default;  // Copy constructor
        Animal(Animal&&) noexcept = default;  // Move constructor
        
        // Destructor
        virtual ~Animal() = default;
        
        // Deleted functions
        void operator=(const Animal&) = delete;
        
        // Methods
        virtual void speak() const {
            cout << name << " makes a sound" << endl;
        }
        
        virtual void move(int distance) {
            cout << "Moving " << distance << " units" << endl;
        }
        
        constexpr int getAge() const { return age; }
        
        // Const method
        const string& getName() const { return name; }
        
        // Static method
        static int count() { return 0; }
        
        // Mutable member access in const method
        void incrementCallCount() const { ++callCount; }
        
        // Volatile method
        volatile void volatileMethod() volatile {}
    };
    
    // Class inheritance
    class Dog : public Animal {
    private:
        string breed;
        
    public:
        Dog(const string& name, const string& breed) 
            : Animal(name), breed(breed) {}
        
        virtual void speak() const override {
            cout << "Woof! I'm a " << breed << endl;
        }
        
        void fetch() final {
            cout << "Fetching..." << endl;
        }
    };
    
    // Abstract class
    class Shape {
    public:
        virtual ~Shape() = default;
        virtual double area() const = 0;
        virtual double perimeter() const = 0;
    };
    
    // Multiple inheritance
    class Drawable {
    public:
        virtual void draw() = 0;
    };
    
    class Circle : public Shape, public Drawable {
    private:
        double radius;
        
    public:
        Circle(double r) : radius(r) {}
        double area() const override { return radius * radius * 3.14159; }
        double perimeter() const override { return 2 * radius * 3.14159; }
        void draw() override { cout << "Drawing circle" << endl; }
    };
    
    // Template classes
    template <typename T>
    class Container {
    private:
        vector<T> items;
        
    public:
        void add(const T& item) { items.push_back(item); }
        T& get(size_t index) { return items[index]; }
        size_t size() const { return items.size(); }
    };
    
    template <typename T, typename U>
    class Pair {
    public:
        T first;
        U second;
    };
    
    template <typename T>
    using IntPair = Pair<T, int>;
    
    // Template specialization
    template <>
    class Container<bool> {
        // Specialized version
    };
    
    // Functions
    void simpleFunction() {
        cout << "Simple function" << endl;
    }
    
    int add(int a, int b) {
        return a + b;
    }
    
    double calculateArea(double radius) {
        return radius * radius * 3.14159;
    }
    
    // Function with default parameters
    void greet(const string& name = "World") {
        cout << "Hello, " << name << endl;
    }
    
    // Template function
    template <typename T>
    T maximum(T a, T b) {
        return (a > b) ? a : b;
    }
    
    // Variadic template
    template <typename... Args>
    void printAll(Args... args) {
        // Implementation
    }
    
    // Inline function
    inline int square(int x) { return x * x; }
    
    // Constexpr function
    constexpr int factorial(int n) {
        return n <= 1 ? 1 : n * factorial(n - 1);
    }
    
    // Reference parameters
    void swap(int& a, int& b) {
        int temp = a;
        a = b;
        b = temp;
    }
    
    // Pointer parameters
    void modify(int* ptr) {
        if (ptr) *ptr = 42;
    }
    
    // Function pointers
    typedef int (*BinaryOp)(int, int);
    
    // Lambda functions
    auto lambda = [](int x) { return x * 2; };
    auto captureByValue = [x = 10]() { return x; };
    auto captureByRef = [&]() { return x; };
    
    // Operators - Arithmetic
    void arithmeticOps(int a, int b) {
        int result = a + b;
        result = a - b;
        result = a * b;
        result = a / b;
        result = a % b;
        a++;
        b--;
        a += 5;
        b -= 5;
    }
    
    // Operators - Bitwise
    void bitwiseOps(int a, int b) {
        int result = a & b;
        result = a | b;
        result = a ^ b;
        result = ~a;
        result = a << 1;
        result = a >> 1;
        a &= b;
        a |= b;
        a ^= b;
        a <<= 1;
        a >>= 1;
    }
    
    // Operators - Logical
    void logicalOps(bool a, bool b) {
        bool result = a && b;
        result = a || b;
        result = !a;
        result = a == b;
        result = a != b;
        result = a < b;
        result = a > b;
        result = a <= b;
        result = a >= b;
    }
    
    // Pointer operations
    void pointerOps(int* ptr, int& ref) {
        int* newPtr = &ref;  // Address-of operator
        int value = *ptr;    // Dereference operator
        int* arrPtr = new int[10];
        delete[] arrPtr;
    }
    
    // Type casting
    void typeCasting(void* voidPtr) {
        int* intPtr = static_cast<int*>(voidPtr);
        const int* constPtr = const_cast<int*>(intPtr);
        Animal* animal = dynamic_cast<Animal*>(nullptr);
        int value = reinterpret_cast<int>(voidPtr);
    }
    
    // Smart pointers
    void smartPointers() {
        unique_ptr<Animal> uniqueAnimal = make_unique<Dog>("Buddy", "Labrador");
        shared_ptr<Animal> sharedAnimal = make_shared<Dog>("Rex", "German Shepherd");
        weak_ptr<Animal> weakAnimal = sharedAnimal;
    }
    
    // RAII pattern
    class FileManager {
    private:
        FILE* file;
        
    public:
        FileManager(const string& filename) {
            file = fopen(filename.c_str(), "r");
        }
        
        ~FileManager() {
            if (file) fclose(file);
        }
    };
    
    // Control flow
    void controlFlow(int x) {
        // if-else
        if (x > 0) {
            cout << "Positive" << endl;
        } else if (x < 0) {
            cout << "Negative" << endl;
        } else {
            cout << "Zero" << endl;
        }
        
        // switch
        switch (x) {
            case 1:
                cout << "One" << endl;
                break;
            case 2:
            case 3:
                cout << "Two or three" << endl;
                break;
            default:
                cout << "Other" << endl;
        }
        
        // for loop
        for (int i = 0; i < 10; ++i) {
            if (i > 5) break;
            if (i == 3) continue;
            cout << i << endl;
        }
        
        // Range-based for loop
        vector<int> numbers = {1, 2, 3, 4, 5};
        for (int num : numbers) {
            cout << num << endl;
        }
        
        // while loop
        while (true) {
            break;
        }
        
        // do-while loop
        do {
            cout << "At least once" << endl;
        } while (false);
        
        // goto (not recommended)
        // goto label;
        // label:
    }
    
    // Exception handling
    void exceptionHandling() {
        try {
            throw runtime_error("Something went wrong");
        } catch (const runtime_error& e) {
            cerr << e.what() << endl;
        } catch (const exception& e) {
            cerr << "Exception: " << e.what() << endl;
        } catch (...) {
            cerr << "Unknown exception" << endl;
        }
    }
    
    // Noexcept specification
    void noexceptFunction() noexcept {
        // Won't throw
    }
    
    void mayThrowFunction() noexcept(false) {
        // May throw
    }
    
    // Comments with special syntax
    // TODO: Implement this feature
    // FIXME: Fix the bug here
    // NOTE: This is important
    // HACK: Temporary workaround
}

// Using declarations
using MyNamespace::Animal;
using MyNamespace::Dog;
using namespace MyNamespace;

// Main function
int main(int argc, char* argv[]) {
    cout << "Hello, World!" << endl;
    
    Dog myDog("Buddy", "Labrador");
    myDog.speak();
    
    return 0;
}
